### PR TITLE
LCAM-244|rounding full assessment amount fields

### PR DESCRIPTION
**Pull request by venkat980**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/LCAM-244)

rounding full assessment amount fields.

## Checklist

Before you ask people to review this PR:

- [x] Tests should be passing: `./gradlew test`
- [x] Github should not be reporting conflicts; you should have recently run `git rebase main`.
- [x] Avoid mixing whitespace changes with code changes in the same commit. These make diffs harder to read and conflicts more likely.
- [x] You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- [x] You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: crime-means-assessment/src/main/java/uk/gov/justice/laa/crime/meansassessment/builder/MeansAssessmentSectionSummaryBuilder.java [hunk 1]
```diff
@@ -83,9 +88,7 @@ public void buildFullAssessment(ApiGetMeansAssessmentResponse assessmentResponse
         apiFullMeansAssessment.setOtherHousingNote(financialAssessmentDTO.getFullOtherHousingNote());
         apiFullMeansAssessment.setTotalAggregatedExpense(financialAssessmentDTO.getFullTotalAggregatedExpenses());
         apiFullMeansAssessment.setTotalAnnualDisposableIncome(financialAssessmentDTO.getFullTotalAnnualDisposableIncome());
-        if (criteriaEntity.isPresent()) {
-            apiFullMeansAssessment.setThreshold(criteriaEntity.get().getFullThreshold());
-        }
+        criteriaEntity.ifPresent(assessmentCriteriaEntity -> apiFullMeansAssessment.setThreshold(assessmentCriteriaEntity.getFullThreshold()));
```
Comment (Mhowell494) [inline crime-means-assessment/src/main/java/uk/gov/justice/laa/crime/meansassessment/builder/MeansAssessmentSectionSummaryBuilder.java:91]: I feel like the original code was more readable. Consider reverting this or improve the formatting.

File: crime-means-assessment/src/main/java/uk/gov/justice/laa/crime/meansassessment/service/MeansAssessmentService.java [hunk 2]
```diff
@@ -143,27 +144,11 @@ BigDecimal calculateSummariesTotal(final MeansAssessmentRequestDTO requestDTO, f
...
-                Frequency applicantFrequency = assessmentDetail.getApplicantFrequency();
-                if (applicantFrequency != null) {
-                    detailTotal = detailTotal.add(applicationAmount.multiply(BigDecimal.valueOf(applicantFrequency.getWeighting())));
-                }
-            }
-        }
-        return detailTotal;
+    BigDecimal calculateDetailTotal(BigDecimal amount, Frequency frequency) {
+        if (amount != null && frequency != null && !BigDecimal.ZERO.equals(amount)) {
+            return setStandardScale(setStandardScale(BigDecimal.valueOf(frequency.getWeighting()))
```
Comment (Mhowell494) [inline crime-means-assessment/src/main/java/uk/gov/justice/laa/crime/meansassessment/service/MeansAssessmentService.java:149]: I'm not sure we also need to set the scale on the frequency weighting.

File: crime-means-assessment/src/test/java/uk/gov/justice/laa/crime/meansassessment/util/SortUtilsTest.java [hunk 3]
```diff
@@ -16,4 +16,11 @@ public void testSortUtilConstructorIsPrivate() throws NoSuchMethodException {
         Constructor<SortUtils> constructor = SortUtils.class.getDeclaredConstructor();
         assertTrue(Modifier.isPrivate(constructor.getModifiers()));
     }
+
+    @Test
+    public void testSort_whenNullIsPassed_NullIsReturned() {
+        List<String> list = null;
+        SortUtils.sortListWithComparing(list, null, null, null);
+        assertNull(list);
```
Comment (Mhowell494) [inline crime-means-assessment/src/test/java/uk/gov/justice/laa/crime/meansassessment/util/SortUtilsTest.java:24]: Can we make sure we're using AssertJ fluent asserts here e.g. assertThat().isNull(). The majority of our other test classes do this and we should be aiming to maintain consistency.

File: crime-means-assessment/src/test/java/uk/gov/justice/laa/crime/meansassessment/util/RoundingUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Modifier;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class RoundingUtilsTest {
+    @Test
+    public void testRoundingUtilConstructorIsPrivate() throws NoSuchMethodException {
+        assertEquals("There must be only one constructor", 1,
```
Comment (Mhowell494) [inline crime-means-assessment/src/test/java/uk/gov/justice/laa/crime/meansassessment/util/RoundingUtilsTest.java:14]: Can we make sure we're using AssertJ fluent asserts here e.g. assertThat().isEqualTo(). The majority of our other test classes do this and we should be aiming to maintain consistency.

File: crime-means-assessment/src/test/java/uk/gov/justice/laa/crime/meansassessment/util/RoundingUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+                RoundingUtils.class.getDeclaredConstructors().length);
+        Constructor<RoundingUtils> constructor = RoundingUtils.class.getDeclaredConstructor();
+        assertTrue(Modifier.isPrivate(constructor.getModifiers()));
```
Comment (Mhowell494) [inline crime-means-assessment/src/test/java/uk/gov/justice/laa/crime/meansassessment/util/RoundingUtilsTest.java:17]: Is it really worth testing whether a constructor is private? This seems superfluous to me.

**Review by Mhowell494 — CHANGES_REQUESTED** [review]

**Comment by Mhowell494** [comment]
Test coverage is at 93%, this should be trivial to get up to 100.

I do wonder whether a different approach might be more maintainable, for example extending BigDecimal and using a wrapper to ensure everything is at the right scale by default. This would avoid having to call the setScale() method whenever we initialise a new BigDecimal. However, this might be worth raising as a separate ticket, since the original bug has been fixed.

https://stackoverflow.com/a/54309652

File: crime-means-assessment/src/main/java/uk/gov/justice/laa/crime/meansassessment/service/MeansAssessmentService.java (hunk 2 above)
Comment (venkat980) [reply crime-means-assessment/src/main/java/uk/gov/justice/laa/crime/meansassessment/service/MeansAssessmentService.java:149]: Setting a scale on the weighting as well result in 4 decimal points after the multiply, hence I removed it.

File: crime-means-assessment/src/test/java/uk/gov/justice/laa/crime/meansassessment/util/RoundingUtilsTest.java (hunk 5 above)
Comment (venkat980) [reply crime-means-assessment/src/test/java/uk/gov/justice/laa/crime/meansassessment/util/RoundingUtilsTest.java:17]: This was added to improve the test coverage.

**Comment by venkat980** [comment]
> Test coverage is at 93%, this should be trivial to get up to 100.
> 
> I do wonder whether a different approach might be more maintainable, for example extending BigDecimal and using a wrapper to ensure everything is at the right scale by default. This would avoid having to call the setScale() method whenever we initialise a new BigDecimal. However, this might be worth raising as a separate ticket, since the original bug is fixed.
> 
> https://stackoverflow.com/a/54309652

There is two lines of code that was not covered that is in the new util class private constructor. It felt superfluous to write tests around that. Agree on the comment that we should look at setting the scale differently.

**Review by Mhowell494 — APPROVED** [review]

**Pull request merged**